### PR TITLE
fix: auto select RV hangs when the daemon is on an older version of noports + unit tests

### LINKED ISSUE #2752: sshnp without -r @rv_xx hangs
### Describe the bug

With the 5.15.0 version I thought we no longer needed to specify the rv. But I see a hang although in the -v mode I do see the latency reports.

### Steps to reproduce

1. Use sshnp without -r

### Expected behavior

nearest rv is used

### Screenshots

```
╰$ sshnp -f @cconstab -t @ssh_1  -d x64 -v      
INFO|2026-07-05 19:54:01.423752|RelaySelector|Checking latency for RVs: [@rv_am, @rv_eu, @rv_ap, @rv_oc] 
INFO|2026-07-05 19:54:01.423905|RelaySelector|Subscribing to notifications with regex: "discover_response\.sshrvd@rv_am" 
INFO|2026-07-05 19:54:01.424570|RelaySelector|Subscribing to notifications with regex: "discover_response\.sshrvd@rv_eu" 
INFO|2026-07-05 19:54:01.424736|RelaySelector|Subscribing to notifications with regex: "discover_response\.sshrvd@rv_ap" 
INFO|2026-07-05 19:54:01.424877|RelaySelector|Subscribing to notifications with regex: "discover_response\.sshrvd@rv_oc" 
INFO|2026-07-05 19:54:01.949933|AbstractAtKeyEncryption (@cconstab)|Encrypted shared symmetric key for @cconstab not found in local storage 
INFO|2026-07-05 19:54:01.950368|AbstractAtKeyEncryption (@cconstab)|Fetching shared symmetric key for @cconstab from atServer 
INFO|2026-07-05 19:54:01.951197|AbstractAtKeyEncryption (@cconstab)|Encrypted shared symmetric key for @cconstab not found in local storage 
INFO|2026-07-05 19:54:01.951216|AbstractAtKeyEncryption (@cconstab)|Fetching shared symmetric key for @cconstab from atServer 
INFO|2026-07-05 19:54:01.951422|AbstractAtKeyEncryption (@cconstab)|Encrypted shared symmetric key for @cconstab not found in local storage 
INFO|2026-07-05 19:54:01.951438|AbstractAtKeyEncryption (@cconstab)|Fetching shared symmetric key for @cconstab from atServer 
INFO|2026-07-05 19:54:01.951797|AbstractAtKeyEncryption (@cconstab)|Encrypted shared symmetric key for @cconstab not found in local storage 
INFO|2026-07-05 19:54:01.951813|AbstractAtKeyEncryption (@cconstab)|Fetching shared symmetric key for @cconstab from atServer 
INFO|2026-07-05 19:54:03.334666|AbstractAtKeyEncryption (@cconstab)|Retrieved my encrypted copy of shared symmetric key for @rv_eu from atServer - saving to local storage 
INFO|2026-07-05 19:54:03.468540|AbstractAtKeyEncryption (@cconstab)|Retrieved my encrypted copy of shared symmetric key for @rv_ap from atServer - saving to local storage 
INFO|2026-07-05 19:54:03.564243|AbstractAtKeyEncryption (@cconstab)|Retrieved my encrypted copy of shared symmetric key for @rv_am from atServer - saving to local storage 
INFO|2026-07-05 19:54:03.660587|AbstractAtKeyEncryption (@cconstab)|Retrieved my encrypted copy of shared symmetric key for @rv_oc from atServer - saving to local storage 
INFO|2026-07-05 19:54:03.769446|AbstractAtKeyEncryption (@cconstab)|Updating public key locally: update:isEncrypted:false:cached:public:publickey@rv_eu MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAz3jGPfhDhSY8HYvAzjtLs6ueYJJSouCq0LxqOzizddvu5DqdhmdtIsrykbwPm24/e2EihzjPzIHnD5kABu5HVwyyX8RgZlwlT3LqrFZZcm2m2UWLcA8Fwcrm1L0TA8JcpqBJAYCFVY7mY+2JQAhC+MjbjEsSbE1BzRYaMVeKRbdGY4fvNR2k4RswM6MbJebYXFQBlsVClMrorOOah36mt3/76+Qfxb+hnmpTevz5IxlaAa4Qi2MCres7TQC2thWTl/rBKseZnJSJ9aouERTAwdn/5RR8tvpJ0zPAl2SHHkH7WaNtlhhj45GLy9zsvoBYZ3S5mdwFmGwYKyHb8aKjYwIDAQAB
 
INFO|2026-07-05 19:54:03.858421|AbstractAtKeyEncryption (@cconstab)|Updating public key locally: update:isEncrypted:false:cached:public:publickey@rv_ap MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAxtNyMjBb+1r1WuWs4GAhFeahv49SEL8Lo76kMLsjzcyJ62ngX9tXEa3u7cpclNfmV9UEnmMVoR46dX/8yQuOsedauYQZWRPDUBqhuBG5TIP7iLj6CXPW0NXHoCFS4XB5JPfxkJo9M8f2xzaraTkXXhcXhd1gxoV7BV+yEmXUdOf/2kQIibp8oGVMiWq/QWO88Rr8AvZJ9cdACesDfrZWXGiDkcMepUGHEvBH7kUBTX27RdrU8bQTbQNzfg14zrMN1tpf8J7V7N1kKK0PWJS5bUUVLvULs4Qu0lqV+3aDJQmUPTb7it30Si7S7tULTeF+HDVmJzKlxTUhJXXuvxUKEwIDAQAB
 
INFO|2026-07-05 19:54:04.025690|AbstractAtKeyEncryption (@cconstab)|Updating public key locally: update:isEncrypted:false:cached:public:publickey@rv_am MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAobv3bVJ7mMkQRd4JPygpdFTwlIZpcuiyRefE4xt9uKPmNHM+38vzKNjGXWJ6U3nO7XkOYnEgzWao2cmID3unUruaj6LdPNlGizkCnl8RmXJNle6Wruxyb9oBSMxDvTVPBJHLTCe8FgIYPhV1Y0COWoglZTjOBL5WY8/h5rwLlDP5jPOLJl0SfY8Hesb9Ki5Ixc8wLlM/ilL6Ay05B2nfFNlyDu8YuwjogvsWnbZJeAnySQRqfCUwG62pE9OcVcyPeO3eVMop72h7c6T30JdaqtCrVH+QD5ngbfXVy0usFoNtXmVQewakyy5StDrJ3Ok88RJzs1rao8OYAxRxzZJVaQIDAQAB
 
INFO|2026-07-05 19:54:04.174575|AbstractAtKeyEncryption (@cconstab)|Updating public key locally: update:isEncrypted:false:cached:public:publickey@rv_oc MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAh30O+5oV1eVFQqprLwlvvUF6PUgis3dwZJZAGU9UhiPNNuvSJa+XI2CdW0IcGcOnhZRGDNHHEqfda/gKf4/BI/LoqlHcq+bLE3hhPgFbECT//+T1dZgvKHEvAVy7vj7W7CP2bIdzeOghxRYp3THj+zifBVuiwRspByAYuTNP2aruy6Z4KKXvQ2GaD8NWUJFbaMXVtIy5uYzQdz3Km7ot6V8xM+zQVsD4GNjHt3NqC1KzUaVGx3KGbkcNeG5U8yyUyMOzcz28tgo3+M7XJiSJhbajMy3xeA6eLdlWW2w5GEvXv4CyWp7qqhNCZdVGFLSEHzVLAxixVDZ0gXi6ceRiMwIDAQAB
 
INFO|2026-07-05 19:54:05.064205|RelaySelector|Subscribing to notifications with regex: "relay_latency_response.x64.sshnp" 
INFO|2026-07-05 19:54:05.335200|RelaySelector|Fetched latencies for client -> RV: {@rv_am: 72, @rv_eu: 228, @rv_oc: 229, @rv_ap: 270} 
INFO|2026-07-05 19:54:05.634880|AbstractAtKeyEncryption (@cconstab)|Encrypted shared symmetric key for @cconstab not found in local storage 
INFO|2026-07-05 19:54:05.634930|AbstractAtKeyEncryption (@cconstab)|Fetching shared symmetric key for @cconstab from atServer 
INFO|2026-07-05 19:54:05.742729|AbstractAtKeyEncryption (@cconstab)|Retrieved my encrypted copy of shared symmetric key for @ssh_1 from atServer - saving to local storage 
INFO|2026-07-05 19:54:05.853758|AbstractAtKeyEncryption (@cconstab)|Updating public key locally: update:isEncrypted:false:cached:public:publickey@ssh_1 MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAqa0UjyA40+8nAqX/3WB7wuEOZLBvvRGeCu59kMryLoQC2Twghjj0BID2fqqE3UCTnMCAUZyHA7xrcEOelyR4JUmzNpV2wamIxAmv+FC38gFqQoVnPhWOZqc0m0FgrmGZ2BaCLLy5WO9efVAP9InkxXg7p1ZM+ALDhQXLEkbpDPqVlPetC9Xd4fKXGqwINkIxzhGsVv5TJj1vuf4ATTwl+Aps9truZqCu+hpHza8RHOQlHFDALxM3hIcvAyXHVm6CoelxI43g/5DxwOJzEZQgUxsVwcXbMfRyAs2G4vsMZlvzaZD6WLbcOYaAWnldKOG/BJH2Fyxm67pLKtILrduxfQIDAQAB
 
^C%                                                                                                                                                                                     ╭─cconstab@tarial in ~ 
╰$ sshnp -f @cconstab -t @ssh_1  -d x64   

```

### MacOS
╰$ uname -a       
Darwin tarial 25.5.0 Darwin Kernel Version 25.5.0: Mon Apr 27 20:38:56 PDT 2026; root:xnu-12377.121.6~2/RELEASE_ARM64_T6000 arm64
╭─cconstab@tarial in ~ 
╰$ 

```
╰$ sshnp   
Version : 5.15.0 (core: 6.12.0)
    --help                               Print this usage information
-k, --key-file                           Sending atSign's atKeys file if not in ~/.atsign/keys/
-P, --pass-phrase                        Pass Phrase to encrypt/decrypt the password protected atKeys file
                                         (defaults to "")
-f, --from (mandatory)                   Sending (a.k.a. client) atSign
-t, --to                                 Receiving device atSign
                                         (defaults to "")
-d, --device                             Receiving device name. Alphanumeric characters, "-" or "_" allowed, max length 36. First character cannot be "-".
                                         (defaults to "default")
-r, --srvd                               The Relay atSign to use. Omit to auto-select the fastest available relay, or provide a comma-separated list to pick the fastest among them.
-l, --local-port                         client-side local port for the ssh tunnel. If not supplied, we will ask the o/s for a spare port
                                         (defaults to "0")
-i, --identity-file                      Identity file to use for ssh connection
    --identity-passphrase                Passphrase for identity file
-s, --send-ssh-public-key                When true, the ssh public key will be sent to the remote host for use in the ssh session
-o, --local-ssh-options                  Add these commands to the local ssh command
-v, --verbose                            More logging
-u, --remote-user-name                   username to use in the ssh session on the remote host
-U, --tunnel-user-name                   username to use for the initial ssh tunnel
    --root-server                        atDirectory domain. Aliases (for backwards compatibility): --root-domain, --rootDomain
                                         (defaults to "root.atsign.org")
    --remote-sshd-port                   port on which sshd is listening locally on the device host
                                         (defaults to "22")
    --idle-timeout                       number of seconds after which inactive ssh connections will be closed
                                         (defaults to "15")
    --ssh-algorithm                      SSH algorithm to use
                                         [ssh-ed25519 (default), ssh-rsa]
    --add-forwards-to-tunnel             When true, any local forwarding directives provided in--local-ssh-options will be added to the initial tunnel ssh request
    --config-file                        Read args from a config file
                                         Mandatory args are not required if already supplied in the config file
    --list-devices                       List available devices. Alias: --ls
    --[no-]authenticate-client-to-rvd    When false, client will not authenticate itself to rvd. Alias: --ac
                                         (defaults to on)
    --[no-]authenticate-device-to-rvd    When false, device will not authenticate itself to rvd. Alias: --ad
                                         (defaults to on)
    --relay-auth-mode                    The authentication mode to use. "escr" (encrypted signed challenge response) is strongest. Alias: --ram
                                         [payload (default), escr]
    --[no-]encrypt-rvd-traffic           When true, traffic via the socket rendezvous is encrypted, in addition to whatever encryption the traffic already has (e.g. an ssh session).
                                         Alias: --et
                                         (defaults to on)
    --daemon-ping-timeout                Seconds the client should wait for response after pinging a daemon. Alias: --dpt
                                         (defaults to "20")
    --[no-]443                           When true, will request ports (443, 443) from relay
    --ssh-client                         What to use for outbound ssh connections
                                         [openssh (default), dart]
-x, --output-execution-command           Output the command that would be executed, and exit
-q, --quiet                              Minimal logging

Invalid argument(s): from (clientAtSign) is mandatory
╭─cconstab@tarial in ~ 
╰$ 
```

### Were you using an atApplication when the bug was found?

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by srieteja**
Closes https://github.com/atsign-foundation/noports/issues/2752

**- What I did**
- Fixed `RelaySelector.selectBestRelay` hanging when a daemon never responds to `relay_latency_request` (pre-5.15.0 daemons, or any stalled/retrying notify send). 
- Added a bounded `deviceLatencyTimeout` (default 20s) around the whole device-latency fetch, and fall back to client-only latency selection instead of failing.
- Added unit tests for `RelayLatencyChecker.measureLatencies`, `RelaySelector` (timeout/fallback/cancellation behavior), and sshnpd's `relay_latency_request` handler and `handleDiscover` edge cases.
- Bumped `noports_core` to 6.12.1 and synced `apps/admin/admin_api` and `packages/dart/sshnoports` lockfiles.

**- How I did it**
- `deviceLatencyTimeout` now bounds the entire wait for the device's `relay_latency_response`, not just the internal completer (which was only armed after `notify()` returned) — so a daemon that never replies degrades to client-only selection after the timeout instead of hanging indefinitely or leaking a `TimeoutException`.

**- How to verify it**
- `cd packages/dart/noports_core && dart test` — all tests pass, including the new regression test for issue #2752 (`silent device must not surface TimeoutException`).
- `cd packages/dart/noports_core && dart analyze` — no issues.
- the new noports client should work with older daemons that dont support the auto RV feature

**- Description for the changelog**
fix: bound and fall back automatic relay selection when device latency check fails; add unit test coverage for auto-RV-select

### CODE REVIEW

File: packages/dart/noports_core/lib/src/common/relay_selector.dart [hunk 1]
```diff
@@ -21,12 +22,20 @@ class RelaySelector with AtClientBindings {
 
   late final String rvServerListUrl;
 
+  /// Injectable latency measurer. Defaults to [RelayLatencyChecker.measureLatencies].
+  /// Override in tests to avoid real TCP connections.
+  @visibleForTesting
+  final Future<Map<String, int>> Function(Map<String, dynamic>)
+  latencyMeasurer;
```
Comment (gkc) [inline packages/dart/noports_core/lib/src/common/relay_selector.dart]: please fix format of the lines above

**Review by gkc — CHANGES_REQUESTED** [review]
overall LGTM; a few small nits

- I noticed one dart format issue, you may need to run dart format on all of the files you've changed

- I _think_ @visibleForTesting on the constructor parameter is redundant; the field already carries it.

- I found this a bit confusing / difficult to read
```
    final Future<Map<String, int>?> deviceLatencyFuture =
        fetchDeviceLatencies(rvIpMap, timeout: deviceLatencyTimeout)
            .timeout(deviceLatencyTimeout)
            .then<Map<String, int>?>((v) => v)
            .onError((e, _) {
              logger.warning(
                'Device latency check failed ($e); '
                'will select relay on client latency only',
              );
              return null;
            });
```

Mayb something like this would be nicer
```
final deviceLatencyFuture = () async {
  try {
    return await fetchDeviceLatencies(rvIpMap, timeout: deviceLatencyTimeout)
        .timeout(deviceLatencyTimeout);
  } catch (e) {
    logger.warning('Device latency check failed ($e); client-only selection');
    return null;
  }
}();
```

- srvd.handleDiscover throws FormatException on malformed JSON (jsonDecode before the is! Map guard). It's caught upstream, but logged as a shout with a stack trace instead of just a warning or whatever

**Comment by gkc** [comment]
@srieteja  For usability reasons, please ensure, when someone is running in non-verbose mode and they don't supply `-r`, that a progress message is shown - something like `No relay supplied, will find the lowest-latency relay available - this may take up to 25 seconds`

**Comment by srieteja** [comment]
Tested the changes on my local. I ensured that the daemon does not respond to the latency_check request and made the client connect to the daemon.
```
INFO|2026-07-08 16:47:43.275440|RelaySelector|Device latency check failed: TimeoutException after 0:00:20.000000: Future not completed 
WARNING|2026-07-08 16:47:43.275510|RelaySelector|Device did not report relay latencies (older daemon?); selecting relay on client latency only 
INFO|2026-07-08 16:47:43.275549|RelaySelector|Selecting fastest RV (client-only): @rv_ap ```

File: packages/dart/noports_core/lib/src/common/relay_selector.dart (hunk 1 above)
Comment (srieteja) [reply packages/dart/noports_core/lib/src/common/relay_selector.dart]: fixed in [337875a](https://github.com/atsign-foundation/noports/pull/2691/commits/337875a6f2f53fde4909738db1507e08e54885ae)

**Comment by srieteja** [comment]
All the review comments have been addressed through [337875a](https://github.com/atsign-foundation/noports/pull/2691/commits/337875a6f2f53fde4909738db1507e08e54885ae) and [e3b2638](https://github.com/atsign-foundation/noports/pull/2691/commits/e3b26387e338878fdcc175f18610820555784c77)

File: packages/dart/noports_core/lib/src/common/relay_selector.dart [hunk 2]
```diff
@@ -42,7 +49,11 @@ class RelaySelector with AtClientBindings {
   ///
   /// If [rvAtSigns] is provided, it will select the best from that list.
   /// Otherwise, it will fetch the standard relays from [rvServerListUrl].
-  Future<String> selectBestRelay({List<Atsign>? rvAtSigns}) async {
+  Future<String> selectBestRelay({
+    List<Atsign>? rvAtSigns,
+    Duration requestTimeout = const Duration(seconds: 10),
+    Duration deviceLatencyTimeout = const Duration(seconds: 20),
```
Comment (JeremyTubongbanua) [inline packages/dart/noports_core/lib/src/common/relay_selector.dart]: I'm confused on what the difference between the two is.

I assume:

- requestTimeout - the # seconds to wait for. If you get 0 RV responses, then fail
- deviceLatencyTimeout - the # seconds to wait for daemon's response on which RVs are the best for them

If so, I think docs would be great for this function

File: packages/dart/sshnoports/lib/src/create_sshnp.dart [hunk 3]
```diff
@@ -26,6 +26,12 @@ Future<Sshnp> createSshnp(
   // If srvdAtSign is not provided, or is a comma-separated list,
   // auto select the best rv
   if (params.srvdAtSign.isEmpty || params.srvdAtSign.contains(',')) {
+    if (!params.verbose) {
+      stderr.writeln(
+        'No relay supplied, will find the lowest-latency relay available; '
+        'this may take up to 25 seconds',
+      );
+    }
```
Comment (JeremyTubongbanua) [inline packages/dart/sshnoports/lib/src/create_sshnp.dart]: I feel like this should be stated at runtime so that we can use AtsignLogger instead of doing this weird `params.verbose` if check. WDYT?

**Review by JeremyTubongbanua — COMMENTED** [review]
Left some ideas

File: packages/dart/noports_core/lib/src/common/relay_selector.dart (hunk 2 above)
Comment (srieteja) [reply packages/dart/noports_core/lib/src/common/relay_selector.dart]: I agree and i have now renamed the param to `relayIpDiscoveryTimeout` in [5679ec6](https://github.com/atsign-foundation/noports/pull/2691/commits/5679ec6ae84337c53b53178fb85d2df4f7220cc9)

File: packages/dart/sshnoports/lib/src/create_sshnp.dart (hunk 3 above)
Comment (srieteja) [reply packages/dart/sshnoports/lib/src/create_sshnp.dart]: I kind of half agree. I agree that the "if" check is weird and have removed it in [5679ec6](https://github.com/atsign-foundation/noports/pull/2691/commits/5679ec6ae84337c53b53178fb85d2df4f7220cc9). But this statement has to be logged here; otherwise, during runtime, we cannot conclude if the RV was auto-selected or provided by the user.

**Comment by srieteja** [comment]
I tried fixing the CI, but could not yet. I will take a thorough look at them and fix them through a separate branch.
Note: The CI e2e_tests failure is present in trunk and has not been introduced by this branch.

**Comment by gkc** [comment]
> I tried fixing the CI, but could not yet. I will take a thorough look at them and fix them through a separate branch. Note: The CI e2e_tests failure is present in trunk and has not been introduced by this branch.

@JeremyTubongbanua the e2e_all test pack has been failing for a while ... can you investigate why? My concern is that there might have been a real regression a while back that isn't covered by the new npe2e test packs?

File: packages/dart/sshnoports/lib/src/create_sshnp.dart (hunk 3 above)
Comment (gkc) [reply packages/dart/sshnoports/lib/src/create_sshnp.dart]: There is specifically a progress logging thing in npt and sshnp mains which is what you see when `-v` isn't set

**Comment by srieteja** [comment]
> I tried fixing the CI, but could not yet. I will take a thorough look at them and fix them through a separate branch. Note: The CI e2e_tests failure is present in trunk and has not been introduced by this branch.

Tracking the CI failure with this ticket https://github.com/atsign-foundation/noports/issues/2763

File: packages/dart/sshnoports/lib/src/create_sshnp.dart (hunk 3 above)
Comment (srieteja) [reply packages/dart/sshnoports/lib/src/create_sshnp.dart]: The same log is now routed through logProcess(). [82b4c09](https://github.com/atsign-foundation/noports/pull/2691/commits/82b4c0930ad818b320866a99f199766b774a256c)

**Review by JeremyTubongbanua — APPROVED** [review]
lgtm

**Review by gkc — APPROVED** [review]
LGTM thanks @srieteja 

We should bump to use at_onboarding_cli 3.1.1 (awaiting publication right now) before we push the next release as it resolves a different problem that cconstab encountered
